pkg/cmd/util/clientcmd/factory.go

```diff
@@ -182,16 +183,32 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 	}
 
 	w.Object = func() (meta.RESTMapper, runtime.ObjectTyper) {
+
+		defaultMapper := ShortcutExpander{RESTMapper: kubectl.ShortcutExpander{RESTMapper: restMapper}}
+		defaultTyper := api.Scheme
+
 		// Output using whatever version was negotiated in the client cache. The
 		// version we decode with may not be the same as what the server requires.
-		if cfg, err := clients.ClientConfigForVersion(nil); err == nil {
-			cmdApiVersion := unversioned.GroupVersion{}
-			if cfg.GroupVersion != nil {
-				cmdApiVersion = *cfg.GroupVersion
-			}
-			return kubectl.OutputVersionMapper{RESTMapper: mapper, OutputVersions: []unversioned.GroupVersion{cmdApiVersion}}, api.Scheme
+		cfg, err := clients.ClientConfigForVersion(nil)
+		if err != nil {
+			return defaultMapper, defaultTyper
+		}
+
+		cmdApiVersion := unversioned.GroupVersion{}
+		if cfg.GroupVersion != nil {
+			cmdApiVersion = *cfg.GroupVersion
 		}
-		return mapper, api.Scheme
+
+		// at this point we've negotiated and can get the client
+		oclient, err := clients.ClientForVersion(nil)
```

Comment:
> This looks like the other methods aren't doing the right thing.

I know that this version is pre-negotiated and for a discovery lookup the version doesn't matter so I went with congruence to make the association obvious.